### PR TITLE
Boost: refactor Critical CSS: Part 1

### LINKED ISSUE #21725: Boost: Refactor Critical CSS class
Currently, Critical CSS module does so many tasks within [one class](https://github.com/Automattic/jetpack/blob/master/projects/plugins/boost/app/modules/critical-css/class-critical-css.php) including Critical CSS generation requests, handling Backend requests, Proxying CSS requests, frontend stylesheet loading manipulation, etc.

Refactor some of the logic within the Critical CSS class to separate objects. Discussion: #21665
Possible refactor:
1. Move DOM manipulation to a separate class
2. Separate Critical CSS builder/Generator.
3. Separate AMP logic to a compatibility file.
4. Maybe more?

### PR DESCRIPTION
**Pull request by pyronaur**
Refactoring the Critical CSS class without modifying the plugin behavior. 

It's going to take some time to fully refactor Critical CSS, but this is at least a good first step.

Fixes #21725 

### Changes in this pull request:
- The following functionalities are moved to separate classes/files:
    - Critical CSS generation management-related logic.
    - Site path providers.
    - All Critical CSS REST API logic.
    - Critical CSS display in the frontend.
    - Recommendations dismissal and reset.
    - AMP compatibility is moved to a separate file.
- A new base class to handle the options array is added.
- Recommendation admin-ajax calls are converted to REST API calls
- Admin bar compatibility.
- Regeneration admin notices.

#### Jetpack product discussion

pc9hqz-cz-p2

#### Does this pull request change what data or activity we track or use?

No

### Testing instructions:
- Check out this branch - `refactor/boost-critical-css`
- Run `jetpack build plugins/boost`
    - This will install all the dependencies and will build the plugin.
    - (This is needed because a lot of class names/filenames have changed.)
- Run `composer phpcs:lint`
    - Make sure that there are no phpcs errors under `projects/plugins/boost`
- Run all end to end testing as per the instructions in 
    - (Handy commands: `pnpm run env-start && pnpm run tunnel-on` and `pnpm run test-e2e`)
    - Make sure that all the tests pass
- Perform extensive manual testing to make sure the plugin works as expected

### CODE REVIEW

**Comment by thingalon** [comment]
E2E tests are failing: https://github.com/Automattic/jetpack/runs/4758053037?check_suite_focus=true

File: projects/plugins/boost/app/modules/critical-css/DisplayCriticalCSS.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/critical-css/DisplayCriticalCSS.php:11]: Is this comment just here to silence a lint rule? It doesn't add anything.

File: projects/plugins/boost/app/modules/critical-css/rest-api/BoostEndpoint.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/critical-css/rest-api/BoostEndpoint.php:10]: I assume this will get moved out of the `Critical_CSS` module at some point?

File: projects/plugins/boost/app/modules/critical-css/rest-api/NonceProtection.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/critical-css/rest-api/NonceProtection.php:18]: TIL about tagging classes by marking them as implementing empty interfaces. Neato.

**Review by thingalon — DISMISSED** [review]
Wow, there are some nice changes in here. I like how clean the main module class for Critical CSS has become, and I like how it's gaining a real API structure.

I've added some questions and comments in-line.

I've found two issues:

1. The [E2E tests are failing](https://github.com/Automattic/jetpack/runs/4758053037?check_suite_focus=true); both in GitHub and my local working copy.
2. I get an HTTP 500 error when I change themes:

```
[11-Jan-2022 01:26:07 UTC] PHP Fatal error:  Uncaught Error: Class "Automattic\Jetpack_Boost\Modules\Critical_CSS\Regenerate_Admin_Notice" not found in /usr/local/src/jetpack-monorepo/projects/plugins/boost/app/modules/critical-css/CriticalCSS.php:190
Stack trace:
#0 /usr/local/src/jetpack-monorepo/projects/plugins/boost/app/class-jetpack-boost.php(513): Automattic\Jetpack_Boost\Modules\Critical_CSS\CriticalCSS->get_admin_notices()
#1 /usr/local/src/jetpack-monorepo/projects/plugins/boost/app/admin/class-admin.php(284): Automattic\Jetpack_Boost\Jetpack_Boost->get_admin_notices()
#2 /usr/local/src/jetpack-monorepo/projects/plugins/boost/app/admin/class-admin.php(145): Automattic\Jetpack_Boost\Admin\Admin->get_shown_admin_notice_ids()
#3 /var/www/html/wp-includes/class-wp-hook.php(303): Automattic\Jetpack_Boost\Admin\Admin->enqueue_scripts()
#4 /var/www/html/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters()
#5 /var/www/html/wp-includes/plugin.php(470): WP_Hook->do_action()
#6 /var/www/html/wp-admin/admin-header.php(102): do_action()
#7 /var/www/html/wp-admin/admin.php(239): require_once('...')
#8 {main}
  thrown in /usr/local/src/jetpack-monorepo/projects/plugins/boost/app/modules/critical-css/CriticalCSS.php on line 190
[11-Jan-2022 01:26:42 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:27:59 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:28:13 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:28:16 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:28:50 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:30:02 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:30:42 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:31:48 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:32:02 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
[11-Jan-2022 01:32:29 UTC] PHP Warning:  filemtime(): stat failed for /usr/local/src/jetpack-monorepo/projects/packages/jitm/src/../build/index.js in /usr/local/src/jetpack-monorepo/projects/packages/assets/src/class-assets.php on line 374
```

File: projects/plugins/boost/app/lib/class-options.php
Comment (pyronaur) [reply projects/plugins/boost/app/lib/class-options.php:13]: I agree 👍

File: projects/plugins/boost/app/modules/critical-css/rest-api/BoostEndpoint.php
Comment (pyronaur) [reply projects/plugins/boost/app/modules/critical-css/rest-api/BoostEndpoint.php]: Yeah, it should.

File: projects/plugins/boost/app/modules/critical-css/DisplayCriticalCSS.php
Comment (pyronaur) [reply projects/plugins/boost/app/modules/critical-css/DisplayCriticalCSS.php]: This shouldn't be there at all :)

File: projects/plugins/boost/app/modules/class-module.php
Comment (pyronaur) [reply projects/plugins/boost/app/modules/class-module.php:20]: This is still in progress. We're aiming to remove all hooks out of constructors to decouple object instantiation from hooks. For example, instantiating an object 1000 times shouldn't attach to `init` hooks a 1000 times, unless it's something that's happening clearly for that specific reason.

File: projects/plugins/boost/app/lib/class-options.php
Comment (pyronaur) [reply projects/plugins/boost/app/lib/class-options.php:13]: Renamed Options to Collection

**Comment by pyronaur** [comment]
Live footage of me cleaning up a few things with Critical CSS: https://www.youtube.com/watch?v=AbSehcT19u0

----

I think I'm finally nearing completion. I've moved all the tangled mess around, introduced a cleaner way of registering REST API endpoints, and as much as I can - tried to make each of the classes portable.

**What I've learned.**

Adding hooks (actions, filters) in constructors makes those classes rigid. Objects themselves shouldn't care whether they're instantiated 1 time or 5000 times. If I want to, I should be able to create Critical CSS object again and again to access various parts of it, provided I can provide that object with the necessary data.

So a good rule of thumb - **don't introduce side effects in constructors**. It's in the name - construct the object, but perform no instructions, don't attach listeners, don't modify the database, etc.

----

**Todo**

While I've done a lot of heavy lifting so far, there are still a few small things that this PR needs before it can be thoroughly reviewed and merged:

- [ ] Various places in the application want to know which modules we have and which modules are active. The API for that is clunky at the moment. So far I've only moved things around, I haven't improved the ergonomics there.
- [ ] Interface names are confusing they need a bit of thought to encourage better practices in the future. This is a nice to have, not a requirement.
- [ ] Interface classes are scattered around due to the many iterations during the refactor. It's time to look at the file structure and reorganize it to make a little more sense.
- [ ] Address PHPCS and other lints
- [ ] Search for all `@TODO` entries. Either convert them to GitHub tickets or just remove them if they'll just become `wontfix`. Never merge anything with todos in it.
- [ ] Make sure tests run equally well compared to the main branch

File: projects/plugins/boost/app/admin/class-admin.php
Comment (eAdnan007) [inline projects/plugins/boost/app/admin/class-admin.php:106]: I love this!

File: projects/plugins/boost/app/admin/class-admin.php
Comment (eAdnan007) [inline projects/plugins/boost/app/admin/class-admin.php:49]: This needs an update.

**Comment by pyronaur** [comment]
After a chat with @eAdnan007, we decided to change things around a bit further so that it's easier to work towards the cloud features. This is a relatively small quality of life change, but at the same time, it helps clarify quite a bit what's going on and how we should move forward.

I'll do final clean-ups tomorrow to get the branch ready for merging and any subsequent changes will have to go into separate PRs. There's still a lot of work to be done here, but if we don't merge soon it will explode.

File: projects/plugins/boost/app/admin/class-admin.php
Comment (thingalon) [inline projects/plugins/boost/app/admin/class-admin.php:58]: 😲 Is _THAT_ what a constructor does?!

**Review by thingalon — APPROVED** [review]
I made a couple of improvements:
- b7401e5 adds an exception to the "unused variable" rule, to allow any that start with an `_`. That way we won't need linter warts any time an interface implementor doesn't want to use a function parameter specified by the interface.
- 8893c9c ensures 500 errors which occur during critical css generation initialization aren't ignored (one crashed the UI during my testing)

This is beautiful work. I particularly love the interfaces, and how they are implemented and used. Let's get this thing merged.

File: projects/plugins/boost/app/modules/lazy-images/class-lazy-images.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/lazy-images/class-lazy-images.php:49]: Ok I just imported my changes from my fixes branch, and there are two tests failing. It is not keeping the "lazy-images" feature in lock-step between Jetpack and Boost. This line is why -- the Modules used to be able to override the default status behavior.

I started down the road to fix this -- I introduced a new `JetpackDeferredStatus` class which inherited from `Status` and override its `is_enabled` and `update` methods to defer to Jetpack's module status.

But then I wasn't sure about where to go next.

I could see a few solutions, but the least kludgey to me would be:
1. To change `Feature` from an interface to a base class, so that it can have default behaviour,
2. Add a `create_status_manager` method to the `Feature` base class, which returns a new `Status` object, and override it for lazy images to return a new `JetpackDeferredStatus` object instead.
3. Profit

But that's a big enough design change that I didn't want to do it without consulting with @pyronaur first.

Alternately, we could:
- Add a `create_status_manager` method to every feature. That's kind of tiresome, or
- Add the deferral logic to the `Status` class, and have it know a bunch of slugs to defer to Jetpack's status. That feels like it puts the intelligence in the wrong place.
- Add a special case to the `Optimization` class and have _it_ know which Features need special status treatment.

None of those options seem clean to me.

So @pyronaur, where will we add this logic?

**Review by thingalon — CHANGES_REQUESTED** [review]

**Review by thingalon — APPROVED** [review]

File: projects/plugins/boost/app/assets/src/js/stores/modules.ts
Comment (pyronaur) [reply projects/plugins/boost/app/assets/src/js/stores/modules.ts:12]: Optimization is a feature, but a feature doesn't necessarily mean optimization.

This came out with a brainstorming call with @eAdnan007 where we had this idea that Boost has other "features" that aren't optimizations, and so they're hanging in the air somewhere.

I actually think this is going to need a follow-up PR to get rid of the "Module" terminology entirely, at least when it comes to describing Optimizations and Features. I think that distinction makes it a lot easier to reason about what belongs where.

Comment (thingalon) [reply projects/plugins/boost/app/assets/src/js/stores/modules.ts:12]: That seems reasonable to me, and I'm fine towards replacing Module with Feature / Optimization.

File: projects/plugins/boost/app/rest-api/endpoints/Generator_Request.php
Comment (pyronaur) [reply projects/plugins/boost/app/rest-api/endpoints/Generator_Request.php:12]: 👍  That makes sense

File: projects/plugins/boost/app/modules/lazy-images/class-lazy-images.php
Comment (pyronaur) [reply projects/plugins/boost/app/modules/lazy-images/class-lazy-images.php:49]: > I started down the road to fix this -- I created a new JetpackDeferredStatus class which inherits from Status and overrides its is_enabled and update methods to defer to Jetpack's module status.

This is certainly one way of doing things. Although, whenever possible - I think we should try to think how we can do this with composition first (instead of inhertiance)

> But then how do we make some Optimizations use Status, and others use JetpackDeferredStatus?

Not entirely sure either, but I think it's worth exploring adding a status contract where both Jetpack_Status and Boost_Status both subscribe to that contract and become interchangeable. At that point, whoever is instantiating a status can decide which type of status is needed (for example, depending on whether Jetpack is active)

But there's a problem with this approach still - because both plugins can be activated without the other, so if the place of storage changes, there's no easy way to silently sync and migrate the option.

I have to look closer at the lazy images module - but I think maybe that module itself should be able to know whether it's active or not? I'm not sure, but it would be a way to neatly fix this.



> **Alternately, we could:**

Hm. I need to do a little bit of thinking on this, but it seems that if we use any type of inheritance it's automatically going to tie down the parent class.

Thinking from a single-responsibility angle - what's the single responsibility we're missing here? Sounds like we're talking about where to store the status - either in Jetpack options or Boost options. So I think that's the hint of the direction we need to go in.

No other classes should be aware that we have a compatibility layer with Jetpack. Where to place it exactly - I'm not entirely sure. I think it needs a bit of thought as a part of #22542

**Review by thingalon — APPROVED** [review]

**Pull request merged**
